Add default Saved Variable hooks

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61928
* #61927
* #61957
* __->__ #61834

Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

~~Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.~~

Expose a pair of functions to Python users: `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()`.
These functions control the hooks applied to saved tensors: all tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor, e.g.:

```python
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

Varal7 committed:
Add default Saved Variable hooks

Summary: First pass at tackling #58659

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


Summary: First pass at tackling #58659

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


First pass at tackling #58659.
Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed as per #60975.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addresses in the current PR or in a future PR. 

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29792193).

Varal7 committed:
Update on "Add default Saved Variable hooks"


First pass at tackling #58659.
Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed as per #60975.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addresses in the current PR or in a future PR.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


First pass at tackling #58659.
Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed as per #60975.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addresses in the current PR or in a future PR.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


First pass at tackling #58659.
Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed as per #60975.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addresses in the current PR or in a future PR.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -44,4 +44,35 @@ namespace torch { namespace autograd {
     }
   }
+
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
```
Not sure if there's a better way to do this

albanD replied:
So the main issue with this is that if multiple threads try to use your context manager at the same time, there are going to be conflicts where they set the same thing.
Note that this is most likely ok if you raise an error if someone tries to set the hooks while they're already set.

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addressed in the current PR or in a future PR.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addressed in the current PR or in a future PR.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addressed in the current PR or in a future PR.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor. Open to discussion whether this should be addressed in the current PR or in a future PR.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented:
After discussing with @albanD, it appears that a more intuitive API is to use `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()` instead of the context-manager to better reflect the fact that it's global.
In a future PR, we will make this thread local.

soulitzer commented:
Would it be simpler to change it to thread_local in this PR and keep the context manager?

soulitzer commented on torch/csrc/autograd/saved_variable_hooks.h:
```diff
@@ -10,4 +10,10 @@ struct TORCH_API SavedVariableHooks {
   virtual ~SavedVariableHooks() = default;
 };
 
+struct TORCH_API DefaultSavedVariableHooks {
```
How do we intend to extend this to a potential C++ api in the future?
It feels a little weird having this interface define only one function to be shared between the (maybe you intend to add reset_hooks as well?).
I feel like its slightly less useful to have this common class (vs the SavedVariableHooks) case because there's no place in the code where you need the polymorphism. The get default engine should already handle that for you right?

Varal7 replied:
It is a good point that right now, I don't need it for polymorphism.
We can always rewrite this code when we need it.

Varal7 commented:
> Would it be simpler to change it to thread_local in this PR and keep the context manager?

I think the change is a bit technical and deserves a PR in itself. Another motivation is that the current PR is blocking for easily benchmarking the next task on CUDA-enabled machines.

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, [one could define](https://github.com/pytorch/pytorch/pull/61834/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5822) a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, [one could define](https://github.com/pytorch/pytorch/pull/61834/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5822) a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, [one could define](https://github.com/pytorch/pytorch/pull/61834/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5822) a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29792193).

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.
All tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, [one could define](https://github.com/pytorch/pytorch/pull/61834/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5822) a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor.

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

zou3519 reviewed (commented):
I had some questions about the semantics of the hooks. Other than that, the code looks reasonable to me

Varal7 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5763,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
         y.sum().backward()
         self.assertEqual(a.grad, y)
 
+    def test_default_saved_variable_hooks_saved_original(self):
```
This test and the next one are testing that the default hooks are registered, used and unregistered correctly. 
The disjunction over the "saved_original" / "did not save original" reflects the two paths in the code that can lead to hooks being registered: [L58](https://github.com/pytorch/pytorch/pull/61834/files#diff-cc9fba479b5beae06b2eea2e390d17796e0341c5b037a20b5bcaccbb0c341030R58) (saved original) and [L79](https://github.com/pytorch/pytorch/pull/61834/files#diff-cc9fba479b5beae06b2eea2e390d17796e0341c5b037a20b5bcaccbb0c341030R75) (did not save original).
It corresponds to the value of `saved_original_` in `SavedVariable.cpp`.

zou3519 replied:
Thanks for the context. Do you think a comment about that would be good to add into the code for future readers?

Varal7 replied:
Sure, I can add a comment

Varal7 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5763,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+                return torch.load(name)
+
+            try:
+                torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
+                a = torch.ones(5, requires_grad=True)
+                y = a * a
+                self.assertEqual(a, y.grad_fn._saved_self)
+
+                y.sum().backward()
+                self.assertEqual(2 * a, a.grad)
```
Yes your understanding is correct. 

`pack` is run when hooks are registered (call to `register_hooks`), so with the current PR, just before exiting the constructor of `SavedVariable`.
`unpack` is called whenever the SavedVariable gets access, so with `y.grad_fn._saved_self` or during the backward pass when it is needed by the `grad_fn` that owns it.

zou3519 replied:
If you want to have a really specific test that checks the above behavior, here's an idea to test that `pack` gets run when the SavedVariable gets created and that `unpack` gets run e.g. when `y.grad_fn._saved_self` gets called:
- Inside the pack function and unpack functions, we can raise a warning
- we can use python warnings capture to check that the warnings do get raised
```
torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
a = torch.ones(5, requires_grad=True)
with warnings.catch_warnings(record=True) as w:
    y = a * a
    # should raise two warnings from a being saved twice
    assert len(w) == 2
```

Varal7 replied:
Sorry, I don't understand why warnings should be raised here.
`y.grad_fn._saved_self` and ``y.grad_fn._saved_other` will both return a value equal to `a`, but the underlying `SavedVariable`s are two distinct objects, which can have their own hooks.
For instance, see  https://github.com/pytorch/pytorch/pull/60975/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5722 where we only get a factor 3.

zou3519 replied:
Here's the full script:
```
import torch
import warnings

def pack(x):
    warnings.warn("pack")
    return x

def unpack(x):
    warnings.warn("unpack")
    return x

torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
a = torch.ones(5, requires_grad=True)
warnings.simplefilter('always')
y = a * a
```
I think `y = a * a` should create a MulBackward node that has two SavedVariable. To construct each of them, the pack() function was called. So this line should cause the pack() function to get called twice (once for each SavedVariable)  and therefore warn on "pack" twice.

Varal7 replied:
Oh that's neat, thanks!

Varal7 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5763,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+        y = a ** 3
+        s = torch.sum(y)
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: 2 * x, lambda x: x)
```
In summary: default hooks are applied to `SavedTensors` when the `grad_fn` that creates them has been created inside the context of those default hooks.

In that sense, the behavior of `torch.autograd.graph.set_saved_tensors_default_hooks` is very similar to `torch.set_grad_enabled`:

When grad is enabled (resp. when default hooks are set), a tensor that is created in that context will be given a `grad_fn` (resp. a `SavedVariable` that is created in that context will be given the default hooks and  packed according to the default pack hook).

In this case, `pow_backward` is `grad * (exp * self.pow(exp - 1))`.
`g.grad_fn` is a `MulBackward0` and  creates one `SavedVariable`: `g.grad_fn._saved_self`, which stores the value of `grad` (this is the first factor 2 from this test).
(`g.grad_fn._saved_other` is `None` because `grad` does not require gradients)
Then, `g.grad_fn.next_functions[1][0].next_functions[0][0]` is a `PowBackward0` which creates the `SavedVariable` : `g.grad_fn.next_functions[1][0].next_functions[0][0]._saved_self` which stores the value of `a` (second factor 2 of the test).

Note that `y.grad_fn._saved_self` also stores the value of `a` but is not affected by the hooks because `y` has been created outside of the context. 

The tensor `a` in itself does not have hooks. Only saved tensors do.

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

~~Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.~~

Expose a pair of functions to Python users: `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()`.
These functions control the hooks applied to saved tensors: all tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor, e.g.:

```python
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5763,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+            self.assertEqual(6 * 8 * a, a.grad)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
```
Previous PRs are already testing the cases where:
*  `pack`/`unpack` are not functions: https://github.com/pytorch/pytorch/pull/60663/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4811
* `unpack` returns something that is not a tensor: https://github.com/pytorch/pytorch/pull/60975/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5741

I added test cases where `pack` expects more than one argument: https://github.com/pytorch/pytorch/pull/61834/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR5739

zou3519 replied:
Awesome, thanks for pointing that out

Varal7 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -0,0 +1,7 @@
+import torch
+
+def set_saved_tensors_default_hooks(pack_hook, unpack_hook):
+    torch._C._autograd._register_default_hooks(pack_hook, unpack_hook)
+
+def reset_saved_tensors_default_hooks():
+    torch._C._autograd._reset_default_hooks()
```
Yes, my intention is to have a PR with the docstrings for all four functions with a link to a new section of [Autograd mechanics](https://pytorch.org/docs/stable/notes/autograd.html) that describes the semantics of the hooks.

zou3519 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5773,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: 2 * x, lambda x: x)
+            g, = torch.autograd.grad(s, (a, ), create_graph=True)
+            g.sum().backward()
+            # factor 4 because pow_backward is grad * (exp * self.pow(exp - 1))
+            self.assertEqual(6 * 4 * a, a.grad)
```
The first 2 comes from `grad` being saved and the second 2 comes from `self` being saved, is that right?

Varal7 replied:
Yes that's correct

Varal7 replied:
Yes, that's correct.

zou3519 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5773,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+            s = torch.sum(y)
+            g, = torch.autograd.grad(s, (a, ), create_graph=True)
+            g.sum().backward()
+            self.assertEqual(6 * 8 * a, a.grad)
```
The factor of 8 is from... `y` getting saved, `a` getting saved, and a `grad` being saved? (Would be good to include in a comment, otherwise future readers are going to have to think a bit)

Varal7 replied:
Actually, `y` is not saved. This test block is the combination of the 2 above. 
In the first block, `a` is saved in `y.grad_fn._saved_self`.
We went over what happens in the second block in https://github.com/pytorch/pytorch/pull/61834#discussion_r675200759.
In the third block, we combine both, so 4 * 2 = 8.
In that sense, `a` is saved twice.

zou3519 replied:
Oh I see, thanks for walking through that. This would definitely be nice to have a comment in the code here

zou3519 commented on test/test_autograd.py:
```diff
@@ -5761,6 +5773,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+                torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
+                a = torch.ones(5, requires_grad=True)
+                y = a * a
+                self.assertEqual(a, y.grad_fn._saved_self)
```
Is the ability to access SavedVariables on the grad_fn a new feature? Is this a public feature? (judging by the underscore, seems like no).

The reason for the question is that I'm worried about what happens if someone tries to mutate _saved_self or change it. For example,
```
y.grad_fn._saved_self.zero_()
y.grad_fn._saved_self = torch.randn(3, 3)
```

Varal7 replied:
This was added by @soulitzer with #52451. If the user mutates `y.grad_fn._saved_self`, it has the same behavior as mutating directly `a`.

If `a` is a leaf, it will fail early with `RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.`
If not, it will fail when the variable is unpacked, thanks to the version counter logic. (`RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation:`)

zou3519 reviewed (approved):
LGTM. I had a question about what happens if someone tries to mutate the SavedVariable using the Python API, but that seems orthogonal to this PR

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

~~Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.~~

Expose a pair of functions to Python users: `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()`.
These functions control the hooks applied to saved tensors: all tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor, e.g.:

```python
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29792193).

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

~~Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.~~

Expose a pair of functions to Python users: `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()`.
These functions control the hooks applied to saved tensors: all tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor, e.g.:

```python
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29792193).

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

~~Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.~~

Expose a pair of functions to Python users: `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()`.
These functions control the hooks applied to saved tensors: all tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor, e.g.:

```python
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29792193).

Varal7 committed:
Update on "Add default Saved Variable hooks"


Fixes #58659. This PR builds directly on top of [#58512](https://github.com/pytorch/pytorch/issues/58512#issuecomment-870763958).

~~Creates a context manager `with torch.autograd.graph.saved_tensors_default_hooks(pack, unpack)` that can be used on the Python side.~~

Expose a pair of functions to Python users: `torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)` and `torch.autograd.graph.reset_saved_tensors_default_hooks()`.
These functions control the hooks applied to saved tensors: all tensors *saved* in that context will be packed using the `pack` function, then unpacked accordingly when needed.

Currently, this works by simply calling `register_hooks` (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a `pack` function that saves to disk whenever the tensor size is too big and returns a filename, then `unpack` simply reads the content of the file and outputs a tensor, e.g.:

```python
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Differential Revision: [D29792193](https://our.internmc.facebook.com/intern/diff/D29792193)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29792193).
